### PR TITLE
If encrypt not specified, escalate if required by default

### PR DESCRIPTION
**Pull request by ciaranj**
Changes the behaviour of the pre-login to automatically try again if the server requires
encryption to be enabled but the user hasn't configured the client in that way.

Still allows for the user to explicitly state that they do not ever want to use
encryption by specifying the value as false.  Or to always use and require encryption by setting the value as true.

This commit partially undoes the work done in ceaa8dc15994fe600b01c5e0d84949f3be0984a0 and is a  change in the behaviour, but brings this driver more inline with other TDS drivers such as jTDS.

### CODE REVIEW

**Comment by ciaranj** [comment]
That thing where you've not worked on a project for so long you discover you're no longer a maintainer ;) doh!

**Comment by v-suhame** [comment]
@ciaranj Did you get to test if retry works? Just noticed that driver sends [`ENCRYPT_NOT_SUP`](https://github.com/tediousjs/tedious/blob/master/src/prelogin-payload.js#L107) instead of `ENCRYPT_OFF`, if encrypt is not enabled. 
As per TDS doc (in 2.26.5 Prelogin section), for client with `ENCRYPT_NOT_SUP` and server with `ENCRYPT_ON` combination, server response is `ENCRYPT_REQ (connection terminated)`. Curious if this impacts retry.
It would be good if driver could send `ENCRYPT_OFF` in first attempt and then `ENCRYPT_ON` on receiving `ENCRYPT_REQ`, it would make it inline with TDS spec.

**Comment by ciaranj** [comment]
Retry does work definitely.  However the server does have errors in the event view, I wonder if that's related :)

**Comment by v-suhame** [comment]
Most likely they are related :) Can you try sending `ENCRYPT_OFF` in prelogin? Hopefully that will clear the error. I don't foresee any side effects if we make this switch from `ENCRYPT_NOT_SUP`.

File: src/connection.js [hunk 1]
```diff
@@ -715,7 +715,9 @@ class Connection extends EventEmitter {
   socketClose() {
     this.debug.log('connection to ' + this.config.server + ':' + this.config.options.port + ' closed');
     if (this.state === this.STATE.REROUTING) {
-      this.debug.log('Rerouting to ' + this.routingData.server + ':' + this.routingData.port);
+      if (typeof this.routingData !== 'undefined') {
```
Comment (arthurschreiber) [inline src/connection.js]: This change here seems unrelated? 🤔 If it's not related to the other changes, could you revert this?

File: src/connection.js [hunk 2]
```diff
@@ -69,7 +69,7 @@ class Connection extends EventEmitter {
         },
         enableArithAbort: false,
         enableAnsiNullDefault: true,
-        encrypt: false,
+        encrypt: undefined,
```
Comment (arthurschreiber) [inline src/connection.js]: 👍 for sane defaults!

File: src/connection.js (hunk 1 above)
Comment (arthurschreiber) [reply src/connection.js]: Ah, no, I see that you're re-using the re-routing logic here. I'm not sure how I feel about this, but if that's working, why not? 👍

File: src/connection.js [hunk 3]
```diff
@@ -753,12 +755,15 @@ class Connection extends EventEmitter {
     });
 
     if (preloginPayload.encryptionString === 'ON' || preloginPayload.encryptionString === 'REQ') {
-      if (!this.config.options.encrypt) {
+      if (this.config.options.encrypt === false) {
         this.emit('connect', ConnectionError("Server requires encryption, set 'encrypt' config option to true.", 'EENCRYPT'));
         return this.close();
+      } else if (typeof this.config.options.encrypt === 'undefined') {
```
Comment (arthurschreiber) [inline src/connection.js]: What do you think about changing the conditions here to be a bit more explicit?

```js
if (this.config.options.encrypt === true) {
  ...
} else if (this.config.options.encrypt === false) {
  ...
} else {
  ...
}
```

**Comment by arthurschreiber** [comment]
Also, as @v-suhame already suggested, can you change the prelogin payload to send `ENCRYPT_OFF`?

File: src/connection.js (hunk 1 above)
Comment (ciaranj) [reply src/connection.js]: Yeah I had the same feeling if I'm honest, but I traded that against the scale of change I'd need to do to encapsulate a 'reconnect' state change :/

File: src/connection.js (hunk 3 above)
Comment (ciaranj) [reply src/connection.js]: I don't mind tbh, from what I recall (this was a while ago!) Semantically I guess it would still behave the same?

**Comment by ciaranj** [comment]
Yeah, it's on my list :/ ( to change the payload)

**Comment by arthurschreiber** [comment]
Do you think just changing the default for the `encrypt` option to `true` would be a simpler and more straightforward fix for this? I don't really see any negative effect of doing this, and I even believe we could do this without bumping the major version. 🤔 

@v-suhame What do you think about this ☝️ proposal? It'd solve the same issue as the code in this pull request, but without adding more code to Tedious.

**Comment by ciaranj** [comment]
Sure thing, it was only done this way so as to not negatively impact any existing consumers of the driver (but you’re right we could rely on semver for that)

**Comment by v-suhame** [comment]
I was on same page as @ciaranj regarding backward compatibility. After some reading, figured its version older than SQL Server 2000 that doesn't support SSL, that hopefully has a very small customer base(or none at all).
It's a thumbs-up for making encrypt true by default.

**Comment by arthurschreiber** [comment]
This will be solved by #694, as we'll default to an encrypted connection instead. #694 will be part of the next semver-major release of `tedious`.
